### DISCUSSION TITLE
Proper way to define a reusable scalar value?

### DISCUSSION THREAD

**Discussion by hjoukl** [pydantic/pydantic#2677] [Question]
I'm just getting my feet wet with pydantic (through the wonderful FastAPI) and it's awesome - thanks a lot! :-)

However I'm wondering if there's any recommended way to define a "scalar" value, i.e. some single field, as opposed to a data structure model - but with full title, description, example, ... metadata.

I can come up with this

```
class LKNRModel(BaseModel):
    """Landkreisnummer"""  # district number
    __root__: str
     
    class Config:
        schema_extra = {
            'example': '08111'
            }
     
    # syntactic sugar to retrieve the __root__ field value
    def __call__(self):
        return self.__root__
 ```

but I'm quite unsure if this is an intended way of doing such a thing. I'm not even sure if accessing `__root__` on the instance is a supported protocol or abuse of private internals...
If a bit clunky this kinda works, allthough you cannot use it as a drop-in str (albeit the __call__ helper to get at the actual value).

Maybe there's a better way of defining simple fields (which are usually just builtin types, probably) and adding the "data model metadata" to 'em? If so, I haven't yet found it (I've seen the discussion in a similar but not quite the same vein here: https://github.com/samuelcolvin/pydantic/issues/1010).

Background:

I'd like to reuse a simple defined field in both multiple models *and* as an OpenAPI parameter value, in my case a "path parameter".

So that in effect it should look like this in an OpenAPI-JSON:
```
  "paths": {
    "/incidences/{lknr}": {
      "get": {
        ...
        "parameters": [
          {
            "required": true,
            "schema": {
              "$ref": "#/components/schemas/LKNRModel"
            },
            "name": "lknr",
            "in": "path"
          }
        ],

      "IncidencesModel": {
        "title": "IncidencesModel",
        "required": [
          "LKNR",
          "Landkreis",
          "Anzahl",
          "Inzidenz"
        ],   
        "type": "object",
        "properties": {
          "LKNR": {
            "$ref": "#/components/schemas/LKNRModel"
          },
      ... 
      },
      "LKNRModel": {
        "title": "LKNRModel",
        "type": "string",
        "description": "Landkreisnummer",
        "example": "08111"
      },     
```

Both the path parameter and the response data model make use of the same simple data field model i.e. they reference `"#/components/schemas/LKNRModel`.

This seems not to be possible with current FastAPI - but I can achieve it with a fairly simple patch (to FastAPI) if I define my models like this:

```
# resource schemas                           
class LKNRModel(BaseModel):                  
    """Landkreisnummer"""  # district number                                   
    __root__: str                            
                                             
    class Config:                            
        schema_extra = {                     
            'example': '08111'               
            }                                
                                             
    def __call__(self):                      
        return self.__root__                 
                                             
                                             
class IncidencesModel(BaseModel):            
    #LKNR: str = Field(description='Landkreisnummer', example='08111')
    LKNR: LKNRModel 
   ...

# main app & routes                        
app = FastAPI(title='Covinfo-Service', version='0.1.0')
  
@app.get('/incidences/{lknr}', response_model=IncidencesModel)
def get_incidences_by_id(
        lknr: LKNRModel,
        db: database.Session = Depends(database.get_db)):
    """Retrieve a district incidence data record.
    """
    incidence = database.read_incidences_by_id(db, lknr=lknr())
    ...
                   
 ```

The simple patch that would be needed for FastAPI is to its `utils.is_scalar_field` method:
```
*** .venv/fastapi-venv/lib/python3.6/site-packages/fastapi/dependencies/utils.py	2021-04-15 14:28:20.736542000 +0200
--- fastapi-patch/utils.py	2021-04-15 14:26:01.089771000 +0200
***************
*** 213,223 ****
  
  def is_scalar_field(field: ModelField) -> bool:
      field_info = field.field_info
      if not (
          field.shape == SHAPE_SINGLETON
!         and not lenient_issubclass(field.type_, BaseModel)
          and not lenient_issubclass(field.type_, sequence_types + (dict,))
          and not isinstance(field_info, params.Body)
      ):
          return False
      if field.sub_fields:
--- 213,225 ----
  
  def is_scalar_field(field: ModelField) -> bool:
      field_info = field.field_info
      if not (
          field.shape == SHAPE_SINGLETON
!         and (not lenient_issubclass(field.type_, BaseModel)
!              or field.type_.__custom_root_type__
!                 and is_scalar_field(field.type_.__fields__['__root__']))
          and not lenient_issubclass(field.type_, sequence_types + (dict,))
          and not isinstance(field_info, params.Body)
      ):
          return False
      if field.sub_fields:
```

Again, I'm unsure if I'm touching private attributes here or following "supported protocol"  in accessing `__custom_root_type__` and `__fields__['__root__']` of a BaseModel-derived class.
If that's actually proper use I'd of course like to suggest the patch to FastAPI...

Any hints greatly appreciated.

Best regards, Holger

**Comment by PrettyWood** [comment]
Hi @hjoukl 
Setting
```py
LKNRModel = Annotated[str, Field(description='Landkreisnummer', example='08111')]
```
wouldn't be enough?

**Reply by hjoukl** [reply]
Hi @PrettyWood , thanks a lot for your answer - this might very well work but I'm currently stuck on Python 3.6.
I obviously didn't notice that information: https://pydantic-docs.helpmanual.io/usage/types/#annotated-types.

Will try to fire up some Python 3.9 installation and check...

**Reply by hjoukl** [reply]
So I got round to try it with Py3.9. While the proposed solution (thanks @PrettyWood)  does run without errors it does not produce an OpenAPI "component schema" LKNRModel. Essentially it inlines the type definition (instead of producing a "named scalar type"):
```
    "/incidences/{lknr}": {
      "get": {
        ...
        "parameters": [
          {
            "required": true,
            "schema": {
              "title": "Lknr",
              "type": "string"
            },
            "name": "lknr",
            "in": "path"
          }
        ],
....
  "components": {
    "schemas": {
....
      "IncidencesModel": {
        "title": "IncidencesModel",
        "required": [
          "LKNR",
...
        ],
        "type": "object",
        "properties": {
          "LKNR": {
            "title": "Lknr",
            "type": "string",
            "description": "Landkreisnummer",
            "example": "08111"
          },
...
```

Also note how the metadata (description and example) is used in the Incidences property "LKNR", but not for the path parameter "lknr". I guess that part is FastAPI's workings rather than pydantic's, as you'd usually use the FastAPI Path() function to add metadata to path parameters.

So not quite there yet...

Sidenote: I just stumbled about https://pydantic-docs.helpmanual.io/install/ and realized that I could just have installed typing-extensions for using Py3.6.
